src/net/tls.cc

```diff
@@ -1113,11 +1113,21 @@ class session : public enable_lw_shared_from_this<session> {
                     // version requested by the client is not supported, the
                     // "protocol_version" alert is sent.
                     auto alert = gnutls_alert_description_t(gnutls_error_to_alert(res, NULL));
-                    return send_alert(GNUTLS_AL_FATAL, alert).then_wrapped([res] (future<> f) {
-                        // Return to the caller the original handshake error.
-                        // If send_alert() *also* failed, ignore that.
-                        f.ignore_ready_future();
-                        return make_exception_future<>(std::system_error(res, glts_errorc));
+                    return handle_output_error(res).then_wrapped([this, res, alert = std::move(alert)] (future<> f) {
+                        std::exception_ptr ex;
+                        // output error future is expected to fail
+                        if (f.failed()) {
+                            ex = f.get_exception();
+                        } else {
+                            on_internal_error_noexcept(tlslog, "expected output error after do_handshake error");
```

Comment:
seastar_logger should be sufficient for something that's expected never to happen.

Reply:
> handle_output_error always returns an exception, so this case cannot happen

Why?  At this point we know only that `gnutls_handshake` returned an error.
Can we be sure that `_output_pending` (that is returned by `wait_for_output()`) contains an error?

Reply:
> I wonder if it isn't easier not to even bother unpacking f (which you know is always a failed future) and we can just return f itself? I think it will simplify this code a lot.

Can do, although I think the internal error case is still theoretically possible, if we have a bug.

Reply:
> > handle_output_error always returns an exception, so this case cannot happen
> 
> Why? At this point we know only that `gnutls_handshake` returned an error. Can we be sure that `_output_pending` (that is returned by `wait_for_output()`) contains an error?

Maybe I'm missing something, but it seems to me that handle_output_error() unconditionally returns an exception - if wait_for_output() succeeded, it has `return make_exception_future(_errror)`, otherwise it has `std::throw_with_nested()`. Isn't this an exception in any case?

Reply:
Handle output error already does everything this extra continuation does, minus the `on_internal_error_noexcept`, which I don't think is appropriate, because `default` is not same as "IO error".

Reply:
Looking at it again, I concur.